### PR TITLE
[clang-cl] Add support for [[msvc::constexpr]] C++11 attribute

### PR DESCRIPTION
**Pull request by ridzenis**
As per agreement with @AaronBallman and @erichkeane, I am re-opening this patch here. Migrated from differential revision: https://reviews.llvm.org/D134475

The current version of the patch has undergone numerous revisions before it became the version I am sending to GitHub.

Below you can find a summary of the history of changes.
The full history of all comments (excluding code-inline comments) is available in this file:
[msvc-constexpr-D134475-comments.md](https://github.com/llvm/llvm-project/files/13257930/msvc-constexpr-D134475-comments.md)

---

Initially I submitted this patch to make `[[msvc::constexpr]]` visible in the AST, with no semantics, and no tests at all. @AaronBallman has told me that you typically don't accept such patches, so I started figuring out semantics of `[[msvc::constexpr]]`.

At first I tried to kind of reverse-engineer it myself, I made several observations, some of them were incorrect.

Later, @cpplearner noted:
> It appears that `[[msvc::constexpr]]` does not make a function `constexpr`, but if `[[msvc::constexpr]]` is used in a function definition and in a call to that function, then the annotated function call can be evaluated during constant evaluation: https://godbolt.org/z/3MPTsz6Yn
>
> Apparently this is used to implement `constexpr std::construct_at`, which needs to call placement operator new, but the latter is not `constexpr`.

Additionally I asked MSVC team to comment about `[[msvc::constexpr]]`. Cody Miller from Microsoft has shared semantics of `[[msvc::constexpr]]` [here](https://developercommunity.visualstudio.com/t/msvc::constexpr-specification/10259132).

Further testing and experiments revealed that, unfortunately not all of his points were correct.

After some trials and errors, I came up with the following definition of semantics for `[[msvc::constexpr]]`

    The `[[msvc::constexpr]]` attribute can be applied only to a function definition or a `return` statement.
    It does not impact function declarations.
    A `[[msvc::constexpr]]` function cannot be `constexpr` or `consteval`.
    A `[[msvc::constexpr]]` function is treated in the same way as a `constexpr` function
    if it is evaluated in a constant context of `[[msvc::constexpr]] return` statement.
    Otherwise, it is treated as a regular function.

This doc-note was added to `MSConstexprDocs`.

As per this definition, we need to be able to track whether current invocation of `[[msvc::constexpr]]` function comes from `[[msvc::constexpr]] return` statement, which in turn comes from a constant evaluation context (i.e. a `constexpr` function).
Thus I went with the approach, which I saw being used in some other cases: a RAII helper (`MSConstexprContextRAII`), and a `bool` flag in `CallStackFrame`.

### CODE REVIEW

**Comment by RIscRIpt** [comment]
Initially I replicated semantics of `[[msvc::constexpr]]` from MSVC, so that it was possible to use it the same way as in MSVC, even `[[msvc::constexpr]] return ::new` from non-std namespace. E.g. https://godbolt.org/z/7eKh5Envz
```cpp
// RUN: %clang_cc1 -fms-compatibility -fms-compatibility-version=19.33 -std=c++20 -ast-dump %s | FileCheck %s

// CHECK: used operator new
// CHECK: MSConstexprAttr 0x{{[0-9a-f]+}} <col:17, col:23>
[[nodiscard]] [[msvc::constexpr]] inline void* __cdecl operator new(decltype(sizeof(void*)), void* p) noexcept { return p; }

// CHECK: used constexpr construct_at
// CHECK: AttributedStmt 0x{{[0-9a-f]+}} <col:46, col:88>
// CHECK-NEXT: MSConstexprAttr 0x{{[0-9a-f]+}} <col:48, col:54>
// CHECK-NEXT: ReturnStmt 0x{{[0-9a-f]+}} <col:66, col:88>
constexpr int* construct_at(int* p, int v) { [[msvc::constexpr]] return ::new (p) int(v); }

constexpr bool check_construct_at() { int x; return *construct_at(&x, 42) == 42; }

static_assert(check_construct_at());
```

However @rsmith raised a concern over changes in `PointerExprEvaluator::VisitCXXNewExpr`:
```diff
   bool IsNothrow = false;
   bool IsPlacement = false;
+  bool IsMSConstexpr = Info.CurrentCall->CanEvalMSConstexpr &&
+                       OperatorNew->hasAttr<MSConstexprAttr>();
  if (OperatorNew->isReservedGlobalPlacementOperator() &&
-     Info.CurrentCall->isStdFunction() && !E->isArray()) {
+     (Info.CurrentCall->isStdFunction() || IsMSConstexpr) && !E->isArray()) {
```
> Do we really need this change? Was our existing check of whether the caller is in namespace std not sufficient for MS' standard library? I'd strongly prefer not to have a documented, user-visible attribute that gives permission to use placement new directly.

If I could choose, I would opt to fully replicate the behavior of MSVC. However, I also acknowledge the concerns that have been raised.

@AaronBallman, @erichkeane, any comments?

**Comment by RIscRIpt** [comment]
Discussion initiated by @AaronBallman of one of important design decisions, which was forgotten (?):

> @AaronBallman:
> It's a good question as to whether we want to support that only when passing `-fms-extensions` or not (it seems like a generally useful attribute); we don't do the same for GNU attributes, but maybe we don't want to follow that pattern? This will be the first attribute we add with the `msvc` vendor namespace.

> @RIscRIpt:
> IMO, this attribute is a clearly Microsoft's extension, thus it should be available only when `-fms-extensions` are enabled.

Note: in the current implementation I enable `[[msvc::constexpr]]` only under `-fms-compatibility -fms-compatibility-version=19.33`, (MSVC 19.33), where this attribute has first appeared.

File: clang/test/SemaCXX/ms-constexpr.cpp
Comment (RIscRIpt) [inline clang/test/SemaCXX/ms-constexpr.cpp:30]: Note: this is a long-term TODO, which is quite difficult to implement with reasonably small changes.

Please let me know, how you typically deal with such cases.

---

Redacted comment from https://reviews.llvm.org/D134475#4288759:

Regarding on absolute matching MSVC implementation, I am not sure we will be able to achieve it with reasonable changes. I am skeptical, because I discovered the following test case: (which you see in the code).

It's a valid code for MSVC; nothing special https://godbolt.org/z/znnaonEhM
However supporting this code seems to be difficult in Clang:

`S2` fails checks of "literal type" in this callstack:
1. [clang::CXXRecordDecl::isLiteral](https://github.com/llvm/llvm-project/blob/d27897ade53ab6f983e442f24880260eed26238a/clang/include/clang/AST/DeclCXX.h#L1425-L1451)
2. [clang::Type::isLiteralType](https://github.com/llvm/llvm-project/blob/d27897ade53ab6f983e442f24880260eed26238a/clang/lib/AST/Type.cpp#L2795)
3. [CheckLiteralType](https://github.com/llvm/llvm-project/blob/d27897ade53ab6f983e442f24880260eed26238a/clang/lib/AST/ExprConstant.cpp#L2392-L2397)

We have information about `CanEvalMSConstexpr` only in `CheckLiteralType`. So, currently, I don't see how we can reasonably check whether `S2` is a valid literal type in context of `[[msvc::constexpr]]`. Two obvious ugly solutions are:
1. Copy all checks from `clang::CXXRecordDecl::isLiteral` to `CheckLiteralType` - two places shall be maintained.
2. Propagate `CanEvalMSConstexpr` down to `clang::CXXRecordDecl::isLiteral`. We could add bool args for both `clang::CXXRecordDecl::isLiteral` and `clang::Type::isLiteralType`. But I don't think it's reasonable for supporting rare vendor-specific attribute. Or is it?

I'll try to dive into this problem again, when I start addressing another round of review comments.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaDeclAttr.cpp:7352]: Is this bit documented?  We allow virtual functions, but only in C++20 or newer mode?  Is there reason for that?

**Review by erichkeane — COMMENTED** [review]
I think this is pretty close, I think skipping on constructors for now is acceptable.  However, i think we should enable the test, and record the diagnostic to ensure we show it is invalid for now (along with better showing the TODO).

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (RIscRIpt) [reply clang/lib/Sema/SemaDeclAttr.cpp:7352]: Prior to C++20 `constexpr` functions were not allowed to be `virtual` [cppreference](https://en.cppreference.com/w/cpp/language/constexpr).
I know this is an attribute, but this matches MSVC behavior: https://godbolt.org/z/snn6c1EW7

File: clang/test/SemaCXX/ms-constexpr.cpp
Comment (RIscRIpt) [reply clang/test/SemaCXX/ms-constexpr.cpp:30]: Created a follow-up #72149.

> However, i think we should enable the test, and record the diagnostic to ensure we show it is invalid for now (along with better showing the TODO).

I hope I understood you correctly: I enabled the test, and added `expected-*` comments alongside with `TODO` mentioning 72149.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (erichkeane) [reply clang/lib/Sema/SemaDeclAttr.cpp:7352]: Got it, thanks for the clarification!

File: clang/test/SemaCXX/ms-constexpr.cpp
Comment (erichkeane) [reply clang/test/SemaCXX/ms-constexpr.cpp:30]: Thanks for the bug report, that helps too!  I haven't seen the patch to enable the test/add the expected lines, but yes, that is what I was asking for.

Comment (RIscRIpt) [reply clang/test/SemaCXX/ms-constexpr.cpp]: I thought I could send send all comments at once (like in GitLab or Phabricator), but unfortunately not. I was in process of rebasing and re-running lit tests.

**Comment by RIscRIpt** [comment]
Oof, sorry, I forgot I should push new changes and rebased changes separately (otherwise "Compare" link does not make sense). If you had checkout `829f8098af96` locally, you can see the diff diff as follows:

```
git range-diff 829f8098af96~2..829f8098af96 7d8b1c0fae42~2..7d8b1c0fae42
```

**Comment by RIscRIpt** [comment]
Rebased onto main, resolved conflicts, re-run lit locally.
Please let me know if I could speed-up review in any way.

**Comment by tru** [comment]
@rnk @amykhuang might be interested in reviewing as well.

File: clang/docs/ReleaseNotes.rst
Comment (erichkeane) [inline clang/docs/ReleaseNotes.rst:221]:
```suggestion
  Since MSVC 19.33 added undocumented attribute ``[[msvc::constexpr]]``, this release adds the attribute as well.
```

File: clang/include/clang/Basic/AttrDocs.td
Comment (erichkeane) [inline clang/include/clang/Basic/AttrDocs.td:3621]:
```suggestion
A ``[[msvc::constexpr]]`` function is treated as if it were a ``constexpr``
```

File: clang/include/clang/Basic/DiagnosticSemaKinds.td
Comment (erichkeane) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:2889]: This diagnostic isn't particularly well worded, and, IMO, would be suited by just adding `virtual` to the select in the diagnostic on 2887.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td
Comment (erichkeane) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:2888]: I don't think we add the `[[` and `]]` to any other diagnostic?  So I think we should jsust say: `attribute 'msvc::constexpr'` everywhere here.

File: clang/docs/ReleaseNotes.rst
Comment (RIscRIpt) [reply clang/docs/ReleaseNotes.rst]: Thanks, applied manually.

File: clang/include/clang/Basic/AttrDocs.td
Comment (RIscRIpt) [reply clang/include/clang/Basic/AttrDocs.td]: Thanks, applied manually. Note, now there are two `if` words nearby. But perhaps it's still better.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td
Comment (RIscRIpt) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: You are right, changed.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td
Comment (RIscRIpt) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: Done.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td
Comment (RIscRIpt) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: Tbh, I don't remember my reasoning. I can only say that MSVC at this moment does not emit any warning/error: https://godbolt.org/z/YfMxYPqEx

Anyway, I changed subjects Stmt to ReturnStmt.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td
Comment (RIscRIpt) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: After the latest changes, this is the only `msvc::constexpr` specific attribute left. What do you think if we use existing diagnostic messages with `constexpr` variant instead of introducing a new diagnostic message?

**Review by ridzenis — COMMENTED** [review]
Addressed review comments; updated my tests to use changed diagnostic messages.

Next I'll upload my branch rebased onto main.

**Comment by erichkeane** [comment]
Also, please don't 'force-push', just push an additional commit to your review.  It erases history and makes reviewing it about 3x harder since history/context gets lost.

File: clang/include/clang/Basic/AttrDocs.td
Comment (AaronBallman) [inline clang/include/clang/Basic/AttrDocs.td:3670]:
```suggestion
Semantics of this attribute are enabled only under MSVC compatibility
```

File: clang/lib/AST/ExprConstant.cpp
Comment (AaronBallman) [inline clang/lib/AST/ExprConstant.cpp:5569]:
```suggestion
    MSConstexprContextRAII ConstexprContext(
```
tiny naming nit.

**Review by AaronBallman — COMMENTED** [review]
I'm really liking the direction here -- this was a much more simple solution than I was expecting. Aside from some tiny nits, this looks good to me, but I leave it to Erich for the final sign-off.

File: clang/docs/ReleaseNotes.rst
Comment (RIscRIpt) [reply clang/docs/ReleaseNotes.rst:239]: Added back.

File: clang/lib/AST/ExprConstant.cpp
Comment (RIscRIpt) [reply clang/lib/AST/ExprConstant.cpp]: Adjusted manually.

File: clang/include/clang/Basic/AttrDocs.td
Comment (RIscRIpt) [reply clang/include/clang/Basic/AttrDocs.td]: Adjusted manually.

**Comment by RIscRIpt** [comment]
> Also, please don't 'force-push', just push an additional commit to your review. It erases history and makes reviewing it about 3x harder since history/context gets lost.

Sorry. I did this, because I am used to such workflow and reviewing [range-diff](https://github.com/llvm/llvm-project/compare/c700d21baf395aafbeea88778d13188650ed0c21..1af977ae49616ee00a64f50302f4cd88f9a51881); but to make it work one have to follow strict rule of pushing changes and rebased branch separately (which I failed to do previous time).

**Review by erichkeane — APPROVED** [review]

File: clang/include/clang/Basic/DiagnosticSemaKinds.td
Comment (RIscRIpt) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: tl;dr afterall I would keep it as-is.

Sorry, I overestimated this possibility.

> attribute 'msvc::constexpr' cannot be applied to the virtual function %1

This could be replaced with existing diagnostic
```
def err_constexpr_virtual : Error<"virtual function cannot be constexpr">;
```

However, two other variants of this new diagnostic:

> attribute 'msvc::constexpr' cannot be applied to the constexpr function %1
> attribute 'msvc::constexpr' cannot be applied to the consteval function %1

Are not easily replaceable. The most similar diagnostic I was able to find was:

```
def err_invalid_decl_spec_combination : Error<
  "cannot combine with previous '%0' declaration specifier">;
```

But, using this diagnostic, warnings are not consistent:

```
/home/richard/projects/llvm-project/clang/test/SemaCXX/ms-constexpr-invalid.cpp:10:3: error: cannot combine with previous 'constexpr' declaration specifier
   10 | [[msvc::constexpr]] constexpr void f1() {}
      |   ^~~~
```

`constexpr` comes after `[[msvc::constexpr]]`, but the message implies that `constexpr` came first.

**Comment by RIscRIpt** [comment]
Thank you!

**Comment by RIscRIpt** [comment]
I don't have write access, but I believe it is ready to be merged.

**Comment by RIscRIpt** [comment]
Suggested commit message:
```
Add support for 'msvc::constexpr' C++11 attribute

This commit adds support for MSVC-specific C++11-style attribute
`[[msvc::constexpr]]`, which was added in MSVC 14.33.
The semantics of this attribute is enabled only under MSVC compatibility
(`-fms-compatibility-version`) 14.33 and higher.
Additionally, default value of `_MSC_VER` was raised to 1433.

The current implementation misses support of:
- `[[msvc::constexpr]]` constructors (see #72149);
  at the time of implementation such support required
  unreasonable amount of changes in Clang.
- `[[msvc::constexpr]] return ::new` (constexpr placement new) from
  non-std namespace (see #74924).

Relevant to: #57696
```

File: clang/test/SemaCXX/ms-constexpr-new.cpp
Comment (Endilll) [inline clang/test/SemaCXX/ms-constexpr-new.cpp:1]: Judging by the failure https://lab.llvm.org/buildbot/#/builders/60/builds/15037, you might want to test for specific triples.

**Review by Endilll — COMMENTED** [review]
Some bots are not happy with one of the tests. I contacted you on Discord about your next steps.

I'll have to revert the patch if I won't hear from you soon.

File: clang/test/SemaCXX/ms-constexpr-new.cpp
Comment (RIscRIpt) [reply clang/test/SemaCXX/ms-constexpr-new.cpp:1]: We can just remove `__cdecl` altogether. I put it there to replicate Microsoft's STL. MSVC accepts that test case without `__cdecl` too.

Regarding other compiler error, it was fixed in 3ec6c72551846b8f4143c8c101a1a6203e85a2aa

Comment (RIscRIpt) [reply clang/test/SemaCXX/ms-constexpr-new.cpp:1]: Opened #74932.

**Pull request merged**
